Update all of the docs to reflect new org service details (#244)
moderneinc/moderne-docs: docs/administrator-documentation/references/repos-csv.md

---
sidebar_label: Creating a repos.csv file
description: How to create a repos.csv file that defines your repositories and, potentially, an organizational hierarchy for them.
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Creating a repos.csv file

Whether you are configuring mass ingestion of repositories into Moderne or you are defining an organizational hierarchy, you will need to create a `repos.csv` file that defines your repositories and configuration around them. 

This doc will provide you with everything you need to know about creating and using this file.

## Supported columns

<details>
<summary>
List of all of the supported columns:
</summary>

> | Column name | Required | Description                                                                                                                                                                                                                                            | Example                                                                      |

Reviewer: cloneUrl,branch is for both
org* is for org hierarchy
everything else is just for mass-ingest

|-------------|----------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------|
| cloneUrl    | `true`   | The URL of the repository that should be ingested.                                                                                                                                                                                                     | `git@github.com:google/guava.git` or `https://github.com/openrewrite/rewrite` |

> | branch      | Required for the agent but not required for mass ingestion  | The branch of the above repository that should be ingested.                                                                                                          | `main`                                                                        |

Reviewer 2: Given that it's shared with DX, I suggest we say "Required for the agent" but instead "Required for the agent (for Moderne Platform admins) or DX (for Moderne DX admins)"

> | branch      | Required for the agent but not required for mass ingestion  | The branch of the above repository that should be ingested.                                                                                                          | `main`                                                                        |

Author: > Given that it's shared with DX, I suggest we say "Required for the agent" but instead "Required for the agent (for Moderne Platform admins) or DX (for Moderne DX admins)"

Sorry I'm a bit unclear what this is asking me to do. Are you talking about the `branch` column or columns in general? I know there's an open request for adding more columns showing agent or dx requirements.

> | branch      | Required for the agent but not required for mass ingestion  | The branch of the above repository that should be ingested.                                                                                                          | `main`                                                                        |

Reviewer 2: I was suggesting that we say "Required for the agent (for Moderne Platform admins) or DX (for Moderne DX admins)" for the `branch` column.

> | branch      | Required for the agent but not required for mass ingestion  | The branch of the above repository that should be ingested.                                                                                                          | `main`                                                                        |

Reviewer: 1. My comment is out of date - when I made it it was referencing only Agent and not DX.
2. The branch column is optional for mass-ingest but not for org hierarchy. Since we're asking the users to use the same file for both, this can be confusing because they might not realize it's _never_ optional when it's the same file used for both mass-ingest and org hierarchy.

> | changeset   | `false`  | If provided, this will check out the repository at this specific commit SHA.                                                                                                                                                                           | `aa5f25ac0031`                                                                |

Author: If change isn't useful for agent or DX or mass ingest then what point does it serve? Should I just remove this entirely?

- A single repository may be associated with multiple organizations by being referenced in multiple rows. In the above
  example, `maven-doxia` is associated with both `Director B` and `Team 1`.
- Repositories can be associated with any organization down the tree, not just the leaves. See line 4 in the `csv` file
  for an example.

:::warning
**Do not** add empty orgs so that the orgs line up with one another.

<Tabs>
<TabItem value="not-this" label="Don't do this">
```bash {showLineNumbers}
cloneUrl,branch,org1,org2,org3,org4
"https://github.com/apache/maven-doxia","master","Team 1","Director A","ALL"
# Non-contiguous series of orgs 👇 
"https://github.com/Netflix/photon","main","Team 2","","Director B","ALL"
```
</TabItem>

<TabItem value="do-this" label="Do this">
```bash {showLineNumbers}
cloneUrl,branch,org1,org2,org3,org4
"https://github.com/apache/maven-doxia","master","Team 1","Director A","ALL"
"https://github.com/Netflix/photon","main","Team 2","Director B","ALL"
"https://github.com/Netflix/ribbon","master","Director A","ALL"
```
</TabItem>
</Tabs>
:::

## Full example

Here is a full example of what the first two rows might look like if you defined all of the columns:

> ```csv

Author: I don't see any recommendation for using quotes left in the docs. Where did you see this? Maybe I'm missing something.

| org*        | `false`  | If you want to configure an organizational hierarchy, you can provide one or more organization columns. Each column will specify an organization the repository should be part of. The column name should be `org` plus a number such as: `org1,org2,org3`.  | `openrewrite`                                                           |
</details>

## Generating this file

To kick-start a `repos.csv` file, we recommend using "[repo fetchers](https://github.com/moderneinc/repository-fetchers)". Repo fetchers are scripts that generate a CSV file of all of your repositories by making API calls to your source code manager.

## Defining an organizational hierarchy

Org hierarchy is best explained with an example. Consider an organization that consists of two teams and two directors:

```
All
├── Director A
|       └── Team 1
|       └── Team 2
├── Director B

```

The following CSV file would represent this organizational structure:

```bash showLineNumbers
cloneUrl,branch,org1,org2,org3

> "https://github.com/apache/maven-doxia","master","Team 1","Director A","ALL"
> "https://github.com/Netflix/photon","main","Team 2","Director A","ALL"
> "https://github.com/Netflix/ribbon","master","Director A","ALL"
> "https://github.com/apache/maven-doxia","master","Director B","ALL"

Reviewer 3: ```suggestion
"https://github.com/apache/maven-doxia","main","Team 1","Director A","ALL"
"https://github.com/Netflix/photon","master","Team 2","Director A","ALL"
"https://github.com/Netflix/ribbon","main","Director A","ALL"
"https://github.com/apache/maven-doxia","main","Director B","ALL"
```

As `main` is no more common than `master`, I would inverse the use of the branches here.

Author: While I appreciate the thought, I am using the branches that are used in these repos. I don't think we need to worry about demonstrating main/master for people. Also, most old code uses master instead of main as people often will not change the branch.

```

In the above file, we define 5 organizations (ALL, Director A, Director B, Team 1, and Team 2) and 3 repositories. Notice that one repository can be defined multiple times if you want two distinct organizations to have access to it.

Also note that organizations on the left are children of organizations on the right. For instance, `Team 1` is a child of `Director A` which is a child of `ALL`.

:::tip
Comma-separated files (CSV) are best manipulated using a dedicated editor such as Microsoft Excel, Google Sheets, or Apple Numbers. Here is what the above CSV would look like on Apple Numbers:

![](./assets/numbers-csv-screenshot.png)
:::

## Formatting details

- A single row may not have more columns than what is defined in the first row.
- A single repository may be associated with multiple organizations by being referenced in multiple rows. In the above
  example, `maven-doxia` is associated with both `Director B` and `Team 1`.
- Repositories can be associated with any organization down the tree, not just the leaves. See line 4 in the `csv` file
  for an example.

:::warning

> Org columns must start immediately after the branch column. Likewise, they must be contiguous. **Do not** add empty orgs
> so that the orgs line up with one another.

Reviewer 2: Most of the warnings you have here are no longer true. You can put the orgs at the end or in the middle. You can also put empty orgs at the beginning or end so that all the orgs line up. The only thing you can't do is have empty orgs in the middle, and even _that_ may work correctly (ignoring the empty ones), I just haven't tested it.

```suggestion
**Do not** add empty orgs in between organizations so that the orgs line up with one another.
```
